### PR TITLE
[VM/VMO BUG] fix a huge allocation and be lazier

### LINKED ISSUE #666: [BUG] Uncaught panic happened in heap_allocator
I run 416.gamess , one of spec CPU 2006 test items,  but a pacnic happened:

```
~ # ./gamess_base.amd64-m64-gcc41-nn < exam29.config 
[ERROR]: Uncaught panic!
panicked at framework/aster-frame/src/vm/heap_allocator.rs:24:5:
Heap allocation error, layout = Layout { size: 659468632, align: 1 (1 << 0) }
```
This pacnic alse happened when I run 434.zeusmp.
the file of 434.zeusmp and 416.gamess: [https://github.com/skpupil/my_spec_test](url)

### PR DESCRIPTION
**Pull request by junyang-zh**
Adding Vmo subrange inheritance is nessasary for current semantics. Without such a fix, userspace programs like #666 would easily break kernel space.

What would happen: an ELF with a size of 11 MiB has a 600+ MiB `.bss` section; our kernel spend an extra space of 1.2+ GB just to load the program (600 MiB in the heap, 600 MiB consisting of commited frames)...

Fix #666 (part of the cause).

### CODE REVIEW

File: kernel/aster-nix/src/vm/vmo/dyn_cap.rs [hunk 1]
```diff
@@ -33,22 +34,25 @@ impl Vmo<Rights> {
     /// so are the access rights.
     pub fn new_slice_child(
         &self,
-        range: Range<usize>,
+        from: Range<usize>,
+        to: Range<usize>,
```
Comment (cchanging) [inline kernel/aster-nix/src/vm/vmo/dyn_cap.rs:38]: I can understand your point, but naming these two ranges as `from` and `to` could easily confuse readers. The meanings of these intervals should respectively represent the range to be inherited within the parent and the range size of the child VMO relative to the parent VMO. The names `from` and `to` might give the initial impression that the range `to` is entirely derived from the `from` interval.

**Review by cchanging — COMMENTED** [review]
Thank you for your contribution! This PR allows segment mapping to delay the allocation of padding frames, which I believe is both correct and necessary. However, this PR introduces a new semantic for the child VMO, in that it allows only a portion of the child VMO to inherit from the parent. I'm not sure if this semantic is needed, because the behavior seems odd to me, and semantically, the same result could be achieved by first inheriting a subinterval of the parent and then resizing it. Moreover, the latter approach seems more in line with the semantics established for segment_vmo.

**Comment by junyang-zh** [comment]
> Thank you for your contribution! This PR allows segment mapping to delay the allocation of padding frames, which I believe is both correct and necessary. However, this PR introduces a new semantic for the child VMO, in that it allows only a portion of the child VMO to inherit from the parent. I'm not sure if this semantic is needed, because the behavior seems odd to me, and semantically, the same result could be achieved by first inheriting a subinterval of the parent and then resizing it. Moreover, the latter approach seems more in line with the semantics established for segment_vmo.

Oh I forgot that we can do resize. Maybe I had a mind fog. Doing that minimizes the change to the current semantics.

**Review by StevenJiang1110 — APPROVED** [review]
LGTM! Thanks for your contribution!

**Review by cchanging — APPROVED** [review]
LGTM! Thanks for your contribution!

**Review by tatetian — APPROVED** [review]
LGTM

**Pull request merged**